### ISSUE TITLE
Using mirage of versions above 0.1.35 fails to build and run

### ISSUE THREAD

**Issue by andSubmarine** [miragejs/miragejs#396]
We are a team that is working on a Nuxt prototype and have been using mirage for this with TypeScript for about 2-4 weeks. Ever since version 0.1.36 was released then node have been unable to build and run our application. The only fix was reverting to an older version of mirage. One error can be seen below:

```
1:34 Module '"miragejs"' has no exported member 'JSONAPISerializer'.
  > 1 | import {Server, Model, Response, JSONAPISerializer} from 'miragejs'
      |                                  ^
```

My question is this: Have the way we should use mirage changed with the new typescript support?  Is there any documentation to describe how we should use mirage with typescript? 

mirage.ts:
```
import {Server, Model, Response, JSONAPISerializer} from 'miragejs'
import users from "./fixtures/users"
import contacts from "./fixtures/contacts"

export function makeServer(environment =  "development") {
    return new Server({
        environment,
        fixtures: {
            users,
            contacts
        },
        models: {
            user: Model,
            contact: Model,
            smsVerification: Model
        },
        serializers: {
            application: JSONAPISerializer
        },
...
```

In the above code, the JSONAPISerializer is not recognized as being available from miragejs and the schema do not seem to contain any of our fixtures (i.e. users for instance is currently an array of json objects). Below is (some of) the packages defined for our node setup: 

package.json:
```
{
...
    "dependencies": {
        "axios": "0.19.2",
        "nuxt": "2.11.0",
        "ts-loader": "6.2.1",
        "typescript": "3.8.3",
        "vue": "2.6.11",
        ...
     },
    "devDependencies": {
        "@babel/core": "7.9.0",
        "@nuxt/typescript-build": "0.6.2",
        "@nuxtjs/eslint-config": "2.0.2",
        "@nuxtjs/eslint-module": "1.1.0",
        "@types/jest": "25.1.4",
        "@vue/cli-plugin-unit-jest": "4.3.0",
        "@vue/cli-service": "4.3.0",
        "@vue/test-utils": "1.0.0-beta.32",
        "babel-core": "7.0.0-bridge.0",
        "babel-eslint": "10.1.0",
        "babel-jest": "25.2.6",
        "eslint": "6.8.0",
        ...
        "jest": "25.2.7",
        "miragejs": "0.1.37",
        "ts-jest": "25.3.1",
        "vue-jest": "3.0.5"
      }
}
```

**Comment by samselikoff** [comment]
There was a bug in .36, the `fixtures` key wasn't typed. That was added and released in .37.

But this is the first time I'm hearing about the Serializer issue. I was under the impression there wasn't any docs needed for the types feature and it would just improve TS devs lives but unfortunately I'm not yet versed in TS so I can't answer the specifics...

@zoltan-nz , any chance you know what's going on with @andSubmarine's team's setup?

**Comment by zoltan-nz** [comment]
@andSubmarine @samselikoff Thanks for reporting the above issue. Type definitions for serializers are missing. I'm just extending `index.d.ts` with it. I keep you guys posted.

**Comment by zoltan-nz** [comment]
@samselikoff are you guys planning to refactor the serializers? It uses `.extend()`, which make it difficult to use in ES6 and in TypeScript.

**Comment by zoltan-nz** [comment]
@andSubmarine could you please test this PR in your project?

You can change temporarily the package version in your `package.json` to

```
        "miragejs": "zoltan-nz/miragejs#master",
``` 

Basic Serializer type definition is added: https://github.com/zoltan-nz/miragejs

Feel free to share your suggestions, so we can improve it further. At this stage the type definition is quite loose.

**Comment by samselikoff** [comment]
Definitely planning to change it, one of the top priorities after we tighten up everything + release a 1.0.

`.extend` is used everywhere (Model.extend, Serializer.extend, Factory.extend and so on). Def needs to go.

**Comment by samselikoff** [comment]
They might not be able to just point at your branch because our npm release has a build step. 

@andSubmarine if you get around to trying it out and put `zoltan-nz/miragejs#master` in your package.json you might have to `cd node_modules/miragejs/` and run `yarn build`.

**Comment by andSubmarine** [comment]
Hi again @samselikoff and @zoltan-nz 

Thanks for the quick responses! It seems that the new PR will fix the compile issues with the serializer. However I still have issues the the fixtures / models / schema. 

mirage.js
```
...
        fixtures: {
            users,
            contacts
        },

        models: {
            user: Model,
            contact: Model,
            smsVerification: Model
        },
       
        routes() {
           ...
          this.get("/auth", function(schema,request) {
           ...
                 const user = schema.users.findBy({ username: json.username });
           ...
           }
           ...

```

The above code produces an error such as the one seen below:
```
37:45 Property 'users' does not exist on type 'Schema<Record<string, { id?: string | undefined; attrs: Record<string, unknown>; modelName: string; save(): void; update<K extends never>(key: K, value: {}[K]): void; update(changes: Partial<{}>): void; destroy():
void; reload(): void; }>>'.
```

It might just be something related to our setup, but it would seem that we will need to use a different configuration of our fixtures in order to use the new version of typescript? Could you provide some insight into this or perhaps give an example of a fixture setup that works in newer versions?

**Comment by zoltan-nz** [comment]
@dfreeman Could you help us a little bit in this case please?

It is not really clear how Models, Factories and Schema work together.
We have a few examples in our tests, but they are quite Ember.js related.
https://github.com/miragejs/miragejs/blob/master/types/tests/factory-test.ts#L4

**Comment by zoltan-nz** [comment]
@andSubmarine Please check out Schema, Factory and Model declarations in these tests files, they might help. https://github.com/miragejs/miragejs/blob/master/types/tests/schema-test.ts#L12

As you probably noticed, Mirage.js does a lot of magic under the hood using the freedom and dynamic nature of the JavaScript. As a result, Mirage.js is clean, nice and easy to use. However, it is not so simple to translate those magics to a strict typed form, unless we separately pre-declare a few types. You right, a clear documentation and guide is still missing about how could we use TypeScript effectively in this library. Type definition inherited heavily from the Ember.js version and there are a few things what we can clean up further, but we should see how the library gonna evolve.

@samselikoff I'm just wondering is there any RFC or description about how the API gonna change and how the v1 will look like, which features gonna be dropped, changed or be more framework independent?

**Comment by dfreeman** [comment]
> We have a few examples in our tests, but they are quite Ember.js related.

What part of those tests is Ember related? I could totally be missing something, but I wasn't under the impression that anything with models, factories, or the schema object were unique to Ember 😄 

> `const user = schema.users.findBy({ username: json.username });`

Unfortunately, this kind of string-manipulation logic ("there's a model called `user`, so the schema has a collection called `users`") isn't representable in TypeScript's type system.

Fortunately, the schema itself also has methods that TypeScript _can_ encode types for, so your lookup above can also be written:

```ts
schema.findBy('user', { username: json.username });
```

Right now, it doesn't look like the `routes` callback to the `Server` constructor actually has the right type info for that call to typecheck, though. I'll see if I can put together a PR to fix that.

**Comment by zoltan-nz** [comment]
@dfreeman Fantastic, thank you! :) You right, not really Ember.js, more like ember-cli-mirage... ;) However, patterns and magics what we use in Serializers or other places highly inspired our favourite framework, isn't it. ;)

And please if you can give us a few examples, how should we use schemas, factories, models, servers, would be awesome. The best would be if we could add a TS version to the main guide samples: https://miragejs.com/docs/getting-started/overview/

**Comment by samselikoff** [comment]
@dfreeman is correct, models/factories/schema are all part of Mirage JS and not specifically related to Ember CLI Mirage.

I regret the `schema.users.*` methods, they were actually inspired by MongoDB rather than Ember, but (just like when Ember Data changed) things would have been cleaner if we stuck with `schema.find(modelName)` methods instead. As Dan points out you actually can do this today but I've never really committed to it as public API because I wanted to rethink it. Guess it always got pushed back...

The reason the `extend` module and the `.extend()` syntax has hung around for so long is because when ES6 modules first came out and we had an opportunity to refactor to them, class fields weren't a think, so the only way to represent

```js
Model.extend({
  users: hasMany()
})
```

would have been something like

```js
class extends Model {
  constructor() {
    super()
    this.users = hasMany()
  }
}
```

which we weren't willing to do. Today of course the landscape looks different so after we ship 1.0 I want to revisit this. The `extend` module is one of the bigger warts in the codebase so is pretty high priority to revisit.

As for a v1.0 of `miragejs`, I don't think we want to make a lot of breaking changes. I think I'd rather solidify the current state of things + address all the outstanding bugs (there are many issues tagged Bug right now but most of them are small issues), ship a 1.0 with the current APIs, then get into a better groove regarding core development which will include new features and potentially breaking changes as we improve the API.

Part of all this was seeing if `miragejs` would have enough traction in the wider JS community to justify us spending time on it, and I think it's proven that it has. So, we're definitely planning on continuing to invest.

So, as far as what that means today for TS users, as well as what it means for v1, I think we should do our best to work within the constraints of the current API. If the `schema.users.all()` methods are the only ones unable to be typed it seems like we should be able to work around that, even if it means exposing the methods Dan suggested above. If we have other problems I'm open to the best ways to go about addressing them.

Happy to add a TS overview but again at this point in time I'm not there in terms of knowledge/understanding to explain. Another hope of mine is once we get back into core development we can start using TS within mirage's codebase itself, I think the project could really benefit from it.

Hopefully that adds a bit of clarity!

**Comment by zoltan-nz** [comment]
@samselikoff Thanks for your update. Totally agree, it makes sense. It is a really great plan.

@dfreeman I try to improve the `ServerConfig` typing. Should it be generic like this:

```
  export interface ServerConfig<R> {
    urlPrefix?: string;
    ...
    routes?: (this: Server, schema?: Schema<R>, request?: Request) => void;
    ...
  }

  export class Server<
    Registry extends Record<string, ModelInstance> = Record<
      string,
      ModelInstance
    >
  > implements ServerConfig {
    constructor(options?: ServerConfig<Registry>);
```

What would you suggest?

**Comment by dfreeman** [comment]
@zoltan-nz making `ServerConfig` generic is one piece of the puzzle, but we also need to relate the type of its `models` and `factories` to its `Registry` type in the context of a `new Server(...)` call.

Unfortunately, because constructors can _only_ accept the type parameters of the class itself, it's not possible to do that without making the `Server` type much more complicated. I've been bouncing some ideas on that back and forth with Sam, and in the meantime I put together #406, which doesn't directly fix this but does unblock having relationships work properly once it _is_ fixed.
